smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/ServiceWaitersGenerator.kt

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package software.amazon.smithy.kotlin.codegen.rendering.waiters
+
+import software.amazon.smithy.kotlin.codegen.KotlinSettings
+import software.amazon.smithy.kotlin.codegen.core.*
+import software.amazon.smithy.kotlin.codegen.integration.KotlinIntegration
+import software.amazon.smithy.kotlin.codegen.model.expectShape
+import software.amazon.smithy.kotlin.codegen.model.getTrait
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.waiters.*
+
+/**
+ * A [KotlinIntegration] that generates the waiters for a service.
+ */
+class ServiceWaitersGenerator : KotlinIntegration {
+    override fun enabledForService(model: Model, settings: KotlinSettings): Boolean =
+        model.operationShapes.any { it.waitableTrait != null }
+
+    override fun writeAdditionalFiles(ctx: CodegenContext, delegator: KotlinDelegator) {
+        delegator.useFileWriter("Waiters.kt", "${ctx.settings.pkg.name}.waiters") { writer ->
+            ctx.allWaiters().forEach(writer::renderWaiter)
+        }
+    }
+}
+
+/**
+ * Gets all the waiters in this [CodegenContext].
+ * @return A list of [WaiterInfo] objects.
+ */
+fun CodegenContext.allWaiters(): List<WaiterInfo> {
+    val service = model.expectShape<ServiceShape>(settings.service)
+
+    fun operationWaiters(op: OperationShape): List<WaiterInfo> =
+        op.waitableTrait?.waiters?.map { (name, waiter) ->
+            WaiterInfo(this, service, op, name, waiter)
+        } ?: listOf()
```

Comment:
[`listOf()` is a synonym of `emptyList()`](https://github.com/JetBrains/kotlin/blob/1b49105c78c566ec42b422cbd46782b47a0d9f40/libraries/stdlib/src/kotlin/collections/Collections.kt#L79-L84). Stylistically, I like `listOf()` because it's the same construct used to build non-empty lists and to me empty lists aren't special—they're just lists with zero elements.

Reply:
TIL

Reply:
I can see your perspective, but I find that `emptyList()` lends readability because it's a better signal of intent over `listOf()`.